File: src/main/java/sk/stuba/pks/service/PacketSender.java
package sk.stuba.pks.service;

import lombok.extern.log4j.Log4j2;
import org.springframework.stereotype.Service;
import sk.stuba.pks.dto.Packet;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;

@Service
@Log4j2
public class PacketSender implements Sender {

    private final DatagramSocket socket;

    public PacketSender(DatagramSocket socket) {
        this.socket = socket;
    }

    @Override
    public void sendPacket(Packet packet, String serverAddress, int serverPort) {
        try {
            byte[] data = packet.getBytes();
            InetAddress address = InetAddress.getByName(serverAddress);
            DatagramPacket datagramPacket = new DatagramPacket(data, data.length, address, serverPort);
            socket.send(datagramPacket);
            log.info("Sent packet: " + packet);
//            System.out.println("Sent packet: " + packet);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: src/main/java/sk/stuba/pks/library/service/PacketSender.kt
package sk.stuba.pks.library.service

import io.ktor.network.sockets.*
import io.ktor.utils.io.*
import io.ktor.utils.io.core.*
import kotlinx.coroutines.delay
import kotlinx.coroutines.withTimeout
import org.reflections.Reflections.log
import sk.stuba.pks.library.dto.Packet
import java.net.BindException
import java.net.SocketException
import java.util.*
import java.util.concurrent.ConcurrentLinkedDeque

class PacketSender(
    private val socket: BoundDatagramSocket,
    private val serverAddress: String,
    private val serverPort: Int,
    private val connectionTimeoutMs: Long,
    private val reconnectEveryMs: Long,
    private val reconnectMaxAttempts: Long,
) {
    private val packetQueue: Deque<Packet> = ConcurrentLinkedDeque()

    suspend fun startSendingPackets() {
        while (true) {
            val packet: Packet? = packetQueue.poll()
            packet?.let {
                sendPacket(it)
            }
        }
    }

    private var sent = 0

    fun addPacket(packet: Packet) {
        packetQueue.add(packet)
    }

    fun addPacketToBeginning(packet: Packet) {
        packetQueue.addFirst(packet)
    }

    suspend fun sendPacket(packet: Packet) {
        sent++
        var isSent = false
        val data = packet.bytes
        var attempts = 0
        withTimeout(connectionTimeoutMs) {
            while (!isSent) {
                try {
                    val addrs = InetSocketAddress(serverAddress, serverPort)
                    val byteReadPacket = ByteReadPacket(data)
                    val datagramPacket = Datagram(byteReadPacket, addrs)
                    socket.send(datagramPacket)
                    isSent = true
                    attempts = 0
                } catch (e: BindException) {
                    if (attempts > reconnectMaxAttempts) {
                        throw CancellationException("Couldn't reconnect")
                    }
                    attempts++
                    log.info("Cant bind, retrying")
                    delay(reconnectEveryMs)
                } catch (e: SocketException) {
                    if (attempts > reconnectMaxAttempts) {
                        throw CancellationException("Couldn't reconnect")
                    }
                    attempts++
                    log.info("Cant bind, retrying")
                    delay(reconnectEveryMs)
                }
            }
            log.trace("Sent packet $sent")
        }
    }

    fun clearQueue() {
        packetQueue.clear()
    }
}
